src/consensus/merkle.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2015-2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <consensus/merkle.h>
+#include <hash.h>
+
+/*     WARNING! If you're reading this because you're learning about crypto
+       and/or designing a new system that will use merkle trees, keep in mind
+       that the following merkle tree algorithm has a serious flaw related to
+       duplicate txids, resulting in a vulnerability (CVE-2012-2459).
+       The reason is that if the number of hashes in the list at a given time
+       is odd, the last one is duplicated before computing the next level (which
+       is unusual in Merkle trees). This results in certain sequences of
+       transactions leading to the same merkle root. For example, these two
+       trees:
+                    A               A
+                  /  \            /   \
+                B     C         B       C
+               / \    |        / \     / \
+              D   E   F       D   E   F   F
+             / \ / \ / \     / \ / \ / \ / \
+             1 2 3 4 5 6     1 2 3 4 5 6 5 6
+       for transaction lists [1,2,3,4,5,6] and [1,2,3,4,5,6,5,6] (where 5 and
+       6 are repeated) result in the same root hash A (because the hash of both
+       of (F) and (F,F) is C).
+       The vulnerability results from being able to send a block with such a
+       transaction list, with the same merkle root, and the same block hash as
+       the original without duplication, resulting in failed validation. If the
+       receiving node proceeds to mark that block as permanently invalid
+       however, it will fail to accept further unmodified (and thus potentially
+       valid) versions of the same block. We defend against this by detecting
+       the case where we would hash two identical hashes at the end of the list
+       together, and treating that identically to the block having an invalid
+       merkle root. Assuming no double-SHA256 collisions, this will detect all
+       known ways of changing the transactions without affecting the merkle
+       root.
+*/
+
+/* This implements a constant-space merkle root/path calculator, limited to 2^32 leaves. */
+static void MerkleComputation(const std::vector<uint256>& leaves, uint256* proot, bool* pmutated, uint32_t branchpos, std::vector<uint256>* pbranch) {
```

Comment:
```suggestion
static void MerkleComputation(const std::vector<uint256>& leaves, uint256* proot, bool* pmutated, uint32_t branchpos, std::vector<uint256>* pbranch)
{
```
not how i wanted to start the review :P hate being the person that does this but we regularly like to keep the code generally in the same formats :)

Reply:
dont worry about all the other place inside the function that does that. it'll be a mass code cleanup one day :)

Reply:
Actually I am fine with the beginning brace being in that position. The Bitcoin style guide actually has it that way, and our original code is actually non-conforming.

I really don't care about

void foo() {
}

versus

void foo()
{
}

Reply:
Personally I preter the "matching brace at the same indent level" because it is easier for the eye to match the brace level that way. I apparently am in the minority.

Reply:
There is one habit we have in the older code that I HATE and we need to clean up when we are going through an area:

if (condition)
    statement;
    
or

if (condition)
{
    statement;
}
else
    statement;
    
I do NOT want people putting single statements after if or for indented without braces, even if it is only one statement. You can put it on the same line as the if or for if it is a single statement, or you can put braces around it. Lots of our code violate this rule and it can cause confusion.

Reply:
On a related note, I suggest that we avoid rewriting ports of Bitcoin code just to change the style. It can cause headaches when we merge upstream code in the future.

Reply:
Yep.